### ISSUE TITLE
Allow plugins to modify existing commands

### ISSUE THREAD

**Issue by fpliger** [pyscript/pyscript-cli#87]
As a plugin author
I'd like to be able to modify an existing command in order to add command options, amend the description, and actually change the command using specific hooks in the command pipeline
so that I can "easily" modify the behaviour a command.

As an example:

I'd like to create a plugin that changes the `create` command so that it can have an additional option `--from-file` so that it can basically do exactly what the `wrap` command does today.

**Comment by fpliger** [comment]
@mchilvers I have an almost working version for this but it's ugly... mostly because of the race condition between the Plugin manager loading the commands and the loading additional options for these very same commands. 

Basically, if commands are defined at the global scope of the module we import them from like it is right now, the commands are already defined when it comes time to load the new options. The quick ugly fix is to define the commands themselves inside the `pyscript_subcommand` hook.

I'll keep you posted

**Comment by mchilvers** [comment]
Ok, so basically the plugins need to register "factories" that create the commands? Could another option might be to make commands callable instances? They could have attributes for arguments/options etc that could be extended by plugins.

**Comment by mchilvers** [comment]
When I say "factory" I mean it in the general sense.... i.e. something that when called produces something... not implying classes etc. 

In out case plugins need to be able to contribute arguments/options to command factories. The factories then produce the commands on-demand for the CLI.

**Comment by fpliger** [comment]
Ok... spoke too soon. Even if I can pass arguments around, I can't think of any possible way to make a command factory work without doing some pretty horrible 🤮 things to make it work. Even if we make it work, it'd also require adding new commands with a really ugly API.

Also, note that I couldn't find anything on the Typer side to support this type of feature. 

I'll let it sit for the day and get back to it with fresh eyes tomorrow

**Comment by mchilvers** [comment]
G'day! Yes - I've been looking at Typer this morning... not sure it will be easy for plugins to contribute arguments etc. This is one of the problems with decorator based APIs... :(

**Comment by fpliger** [comment]
Yeah. Here's where my thoughts are heading. We hold this specific [very] nice feature and focus on adding features that are really must haves for the upcoming release. In the meantime, we start a discussion with the `Typer` folks to explore ways we can do it.
